Create a Tikz latex drawing feature for networkx

dschult opened the pull request:
This PR provides a drawing feature (module and tests) from NetworkX to the TikZ library of TeX/LaTeX.
The basic interface is `nx.to_latex(G, pos, **options)` to construct a string of latex code or 
`nx.write_latex(G, filename, as_document=True, **options)` to write the string to a file. 
The `as_document` option puts the LaTeX figure inside a document environment with preamble
that loads TikZ. When False, the code for the figure stands alone in the file ready to be `\input` into your LaTeX file.

Alternatives to #5639 and #5702  
This makes them obsolete if merged.

MridulS commented on networkx/drawing/nx_latex.py:
```diff
@@ -0,0 +1,570 @@
+        # edge text is drawn using the TikZ node command inside curly brackets {}
+        e_label = f" node{els} {{{edge_label[edge]}}}" if edge in edge_label else ""
+
+        result += i8 + f"\\draw ({u}) to{e_opts} ({v}){e_label};"
```
```suggestion
        result += i8 + f"\\draw{e_opts} ({u}) to ({v}){e_label};"
```

This code doesn't work on my standard installation of LaTeX on a MacOS from https://www.tug.org/mactex/

MridulS reviewed (commented):
Thanks for this @dschult!

We should try to get this in soon, it's a great feature. Interestingly a bunch of the generated LaTeX doesn't compile on my machine. It has a standard MacOS LaTeX distribution from https://www.tug.org/mactex/

I have suggested the changes to make it work with this distribution. Does the LaTeX still compile for you with the changes? I also tried to follow up the format from https://tikz.dev for these changes.

MridulS commented on networkx/drawing/nx_latex.py:
```diff
@@ -0,0 +1,570 @@
+
+If you want **subfigures** each containing one graph, you can input a list of graphs.
+
+>>> H1 = nx.path_graph(4)
+>>> H2 = nx.complete_graph(4)
+>>> H3 = nx.path_graph(8)
+>>> H4 = nx.complete_graph(8)
+>>> graphs = [H1, H2, H3, H4]
+>>> caps = ["Path 4", "Complete graph 4", "Path 8", "Complete graph 8"]
+>>> lbls = ["fig2a", "fig2b", "fig2c", "fig2d"]
+>>> nx.write_latex(graphs, "subfigs.tex", n_rows=2, sub_captions=caps, sub_labels=lbls)
```
https://github.com/networkx/networkx/pull/6238#discussion_r1052010368 should fix it.

dschult commented on networkx/drawing/tests/test_latex.py:
```diff
@@ -0,0 +1,289 @@
+    for aa, bb in zip(expected_tex.split("\n"), output_tex.split("\n")):
+        if aa != bb:
+            content_same = False
+            print(f"-{aa}|\n+{bb}|")
```
This entire block of code is only there to print the parts that are different so that when pytest runs you can see the difference in the results compared to expected. We should either keep it all or remove it all (maybe best to leave it there commented out?  Commented out code eventually gets out of date. But what about code that is helpful for people who are debugging failed tests?  There isn't really a place for that in the normal workflow.).  :}

dschult commented on networkx/drawing/tests/test_latex.py:
```diff
@@ -0,0 +1,289 @@
+    pos = {n: (n, n) for n in G}
+
+    G.nodes[0]["style"] = "blue"
+    G.nodes[2]["style"] = "line width 3,blue!50"
```
I think the line width for a node is the width of the outline of the shape -- e.g. a circle that is not filled will have a line width.  But I have to say, there are no many options with tikz it is hard to keep it all straight.  I'll get my head back into it and then will be able to answer better...

dschult replied:
You are correct:  I can't find any style for nodes or edges called `line width` and the latex code created in the test doesn't compile with latex.  In futzing with that, I found that the edge labels should be placed before the tail node of the edge when using the "to" syntax (as opposed to the "--" syntax).

dschult replied:
OK... Now I'll correct myself... linewidth is n option for both nodes and edges, but the syntax is:  `line width=3` not `line width 3`.  The frustrating part of that is that I was able to find that on StackOverflow but not in the pgf/tikz docs.  I think I have corrected the examples and the tests, but I'll look through it again later to make sure...

dschult commented on networkx/drawing/nx_latex.py:
```diff
@@ -0,0 +1,570 @@
+        pos = nx.get_node_attributes(G, pos)
+    if not pos:
+        # circular layout with radius 1
+        pos = {n: f"({round(2* 3.1415 * i / len(G), 3)}:1)" for i, n in enumerate(G)}
```
Hmmm... What should that default be?  I don't understand why setting it to 1 doesn't work. The distances between positions 1 apart shows up just fine. I guess when using polar coords the result doesn't get scaled in the same way as cartesian?  I haven't checked that.  Anyway, changing it to 10 looks reasonable for these graphs -- probably not for others. But I think this is a reasonable default value.

dschult commented:
Thanks very much @MridulS !! There were some style errors for the edges that this caught nicely. And the default radius for spacing was much smaller than I anticipated (and didn't check). 

I've changed the placement of the edge style code within the `draw __ to __` syntax to allow the edge styles to actually work -- as you suggested. And I also moved the edge label placement in that same construction so the labels are styled and placed correctly.  Hopefully this now works better.

The `line width` style does seem to work with an equals sign `line width=3` for both nodes (outline of the edge shape) and edges (width of the line draw). I hope I've got the tests checking for code that does actually work with Latex.  

I am also using a standard generic install of MacTex (which I think is TexLive plus some mac based GUIs that we're not using).  So any errors you see, I am seeing too (just sometimes not looking the right way :).

I turned off the multigraph for now as suggested. I don't think it would be hard to make it bend differently for the multiedges in a similar way to Dilara's treatment of multiedges in nx_pylab. But that can be a TODO.

Please check it again when you have time.

MridulS reviewed (approved):
Thanks for getting this done @dschult! Maybe we can squeeze this in the 3.0 release.

All looks good to me but even the default radius of 10 for polar coordinates doesn't seem to make the plots look appealing. Maybe we should just leave it to the user for now. We can add more better default layouts in further iterations of this.

The current subfigure test plots this out: 
<img width="761" alt="Screenshot 2023-01-02 at 5 22 44 PM" src="https://user-images.githubusercontent.com/5363860/210257569-17173680-b309-4852-9369-0aba100a15bc.png">

If we change the default radius to 50 we get 
<img width="656" alt="Screenshot 2023-01-02 at 5 24 00 PM" src="https://user-images.githubusercontent.com/5363860/210257625-44057d52-d11e-4cf2-ac29-62eb6ec620a7.png">

But this isn't a blocker for me and I don't know what is the right answer :)

Let's get this in and we can make changes in future PRs!

dschult commented:
Hmmm...  None of these seem to be laid out on a circle...  
I think that's why we can't hardly see any of the edges.  
A linear layout of the nodes isn't a good layout to show the network.
It almost seems like what I think of as polar coordinates are being interpreted as cartesian coordinates.

I'm all for getting this in and then making good examples later.
Thanks!

rossbar reviewed (approved):
Wow this is really nice, thanks @dschult ! I took the liberty of pushing up a few minor docstring/rst formatting nits while reading through. The code for the conversion functions is very nice and straightforward to follow, and the tests seem like just about the best that can be done without some complicated pixel-matching scheme.

I think this is a great new feature and a solid base to expand upon as users try it out.
